File: 20. Valid Parentheses/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const obj = {
    '(': ")",
    '{': '}',
    '[': "]"
  };
  const p = [];

  for (let k = 0; k < s.length; k++) {
    if (obj[s[k]]) {
      p.push(obj[s[k]]);
      continue;
    }

    if (s[k] !== p[p.length - 1]) {
      return false;
    }
    p.pop();
  }

  return (p.length === 0);
};

File: 20. Valid Parentheses/solution.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if s == '':
            return True

        arr = []
        dic = {
            ")": "(",
            "}": "{",
            "]": "["
        }

        for char in s:
            if char in dic.values():
                arr.append(char)
            elif char in dic.keys():
                if not arr or dic[char] != arr.pop():
                    return False
            else:
                return False
        return not arr